### ISSUE TITLE
comments and corrections

### ISSUE THREAD

**Issue by jschmied** [0xBakeer/Qwen3.8-27B-4-bit-on-a-single-DGX-Spark#1]
Some measurements that bear on your DFlash 2 conclusion — and one place where you have data I don't

Thanks for publishing this. The concurrency framing is the part I can't argue with, and I'll come back to it. But I think one conclusion is now out of date, and it's the one that pushed you to llama.cpp.

DFlash 2 does run on a 4-bit checkpoint with a quantized lm_head under vLLM.

The refusal isn't a property of the method — it's an eight-line guard at the top of compute_candidates in qwen3_dflash2.py. The line it protects is:

logits = self.lm_head.quant_method.apply(self.lm_head, hidden_states, bias=None)

quant_method.apply() is the generic interface every scheme implements, and the padding and TP handling below it are scheme-agnostic. Nothing in candidate top-k needs raw weights. Deleting the guard is sufficient - no dequantization, the head stays at its packed size. (Someone made the same argument on vllm#52816 in different words; it hasn't merged.)

Measured on a DGX Spark, 20 real agent contexts, DFlash 2 n=7 on the #52816 branch:

│           target           │           head           │   decode   │ TTFT  │
│ RadixArk/Qwen3.8-27B-NVFP4 │ NVFP4, 4-bit activations │ 40.3 tok/s │ 4.5 s │
│ unsloth/Qwen3.8-27B-NVFP4  │ FP8                      │ 36.5 tok/s │ 5.6 s │
│ our W4A16 build            │ FP8, A16                 │ 41.7 tok/s │ 9.6 s │
  
Thirteen cells in total. The heads that are quantized came out fastest - the five fastest of eleven grid cells all have quantized heads, and the four with untouched BF16 heads are the slowest. That's the opposite of what the guard implies, and it makes sense: under speculation the head runs once per draft position while the body runs once per step, so head bytes dominate.

Your AutoRound observation — bf16 head, but vLLM still assigns a quantized method - is vllm#52883. compute_candidates tested isinstance(quant_method, UnquantizedEmbeddingMethod), and a checkpoint that leaves lm_head alone while carrying quantization_config gets UnquantizedLinearMethod instead. That's fixed; it was a defect, not a property.

Net: 40.3 tok/s on vLLM against the 37.58 you get from llama.cpp, without leaving the stack.

"W4A4 buys nothing because the activation path is not executed in FP4" - right for decode, wrong for prefill.

Across those thirteen cells decode sits between 34.7 and 41.7 with no clear body effect, so I agree there. Prefill is a different story:

W4A16 body            TTFT 9.6 s
W4A4, dynamic act.    TTFT 5.6 s
W4A4, static act.     TTFT 4.5 s

Roughly a factor of two, and static activations are the fastest of all — they skip the per-token scale computation. I'd treated that column as dominated and left it unmeasured for exactly the reason your sentence gives; it turned out to hold the fastest cell on the page.

Whether that matters depends on the workload. For agent and tool-loop traffic it dominates: across my corpus the median turn is 130 completion tokens against a 9,138-token context, so TTFT is 52–75 % of the wall time, and ranking by tok/s  inverts the real ordering. RadixArk decodes 3 % slower than my fastest cell and finishes the turn 39 % sooner.

Where you're ahead of me: concurrency.

Every number above is single-stream. Your c8 → c16 collapse (+10 % → +0.2 %) is data I don't have, and it lines up with vllm#42505 reporting DFlash slower than baseline above ~8 streams. So your headline - quantization pays at low concurrency - is untouched by anything here. If you're serving a fleet rather than one agent, your conclusion stands and mine doesn't apply.

Two smaller things.

You describe VLLM_MARLIN_USE_ATOMIC_ADD=1 as required to avoid a race condition. I have it set, but my notes record it as a throughput flag (+4–8 % at concurrency ≥ 4) on the strength of the load-time log message. If it's a correctness issue, that's worth stating loudly - it changes it from tuning to mandatory, and I'd like to fix my own notes. Do you have a reference?

And I'm going to steal your fresh/edit-heavy split. A 2.5× spread between 29.55 and ~73 tok/s from prompt character alone is larger than most of the effects either of us is measuring, and I report a single median over mixed contexts, which hides it.

Your closing point - that something in this line passed standard benchmarks and then failed a real code task — matches my experience exactly. That's why the unmeasured cells on my map say unknown rather than carrying a derived number.

reference: https://claude.ai/code/artifact/1c0eb53f-9daf-406e-ad6f-149ac0ac161f
